### ISSUE TITLE
Cannot build `mawk` with `clang` on Darwin

### ISSUE THREAD

**Issue by eschnett** [spack/spack#3647]
When I build `mawk` with Apple's `clang` on Darwin 16.5.0 (macOS 10.12.4), then the resulting `mawk` executable always crashes.

A debugger shows the following:
```
$ lldb /Users/eschnett/src/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/mawk-1.3.4-7ktgundnge66kpkge2zyidmx5vq64mjg/bin/mawk
(lldb) target create "/Users/eschnett/src/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/mawk-1.3.4-7ktgundnge66kpkge2zyidmx5vq64mjg/bin/mawk"
Current executable set to '/Users/eschnett/src/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/mawk-1.3.4-7ktgundnge66kpkge2zyidmx5vq64mjg/bin/mawk' (x86_64).
(lldb) r
Process 16120 launched: '/Users/eschnett/src/spack/opt/spack/darwin-sierra-x86_64/clang-8.1.0-apple/mawk-1.3.4-7ktgundnge66kpkge2zyidmx5vq64mjg/bin/mawk' (x86_64)
2017-03-31 08:47:27.589547-0400 mawk[16120:5968363] detected buffer overflow
Process 16120 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00007fffc6d74d42 libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fffc6d74d42 <+10>: jae    0x7fffc6d74d4c            ; <+20>
    0x7fffc6d74d44 <+12>: movq   %rax, %rdi
    0x7fffc6d74d47 <+15>: jmp    0x7fffc6d6dcaf            ; cerror_nocancel
    0x7fffc6d74d4c <+20>: retq
(lldb)
```

Note the "detected buffer overflow" message.

I can build `mawk` just fine with a Spack-built `gcc`.

**Comment by adamjstewart** [comment]
*sigh*. I was hoping this would be a good solution to the gawk bootstrapping problem. Maybe it would be better to remove all dependencies on gawk/mawk and simply patch the things that require them so that they work with any system awk implementation. Otherwise, we'll have to contact the mawk developers to fix this.

**Comment by davydden** [comment]
i confirm that I had the same issue. I don't think `awk` is a virtual dependency currently, maybe it should be? Otherwise one has to set `mawk` as an external package provided by macOS.

**Comment by adamjstewart** [comment]
It is a vdep.

**Comment by eschnett** [comment]
`ncurses` depends on `mawk`, not on the virtual `awk`.

**Comment by adamjstewart** [comment]
Yep, that was just to work around the circular dependency.

**Comment by boegel** [comment]
I'm hitting the same problem.

Things I've tried to work around, none of them worked:

* add back the `CPPFLAGS=-P` option for ncurses (cfr. #3361, was dropped in #3425): same problem remains
* add dependency on `gawk` rather than `mawk` (cfr. #3425, reverting in #3481 to avoid circular dependency): introduces circular dependency (couldn't Spack report this more nicely?)
* update `mawk` to the patched `1.3.4-20161120` version available from ftp://invisible-island.net/mawk/ (the original `1.3.4` dates from Dec'09...): doesn't help
* remove `mawk` dep to rely on the system `awk`: same problem (macOS 10.12.3; `awk --version` spits out `awk version 20070501`)

How does switching to `gawk` introduce a cyclic dependency exactly @adamjstewart?

**Comment by adamjstewart** [comment]
The problem is that in terms of dependencies, `gawk` -> `gettext` -> `ncurses` -> `mawk`. `ncurses` obviously can't depend on `gawk` or else it would become a cycle. Normally, every package in Spack would use the system `awk` installation, but @eschnett found a package that couldn't be built with his system `awk`, so he added a dependency on `gawk`. Then I noticed that `gawk` depended on `gettext`, creating the loop. `mawk` was the solution I came up with at the time, but it might be simpler if we just removed all `awk` dependencies and patched the source code of `ncurses` so that it would build with the system `awk`.
